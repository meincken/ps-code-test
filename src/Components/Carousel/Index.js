import React, { Component } from "react";
import styled from "styled-components";
// import Controls from "./Controls";
// import Item from "./Items";
import { ReactComponent as NavigationIcon } from "../../Icons/arrow.svg";

const Section = styled.section`
  max-width: 90vw;
  margin: 0 auto;
`;

const Article = styled.article`
  display: flex;
  flex-wrap: nowrap;
  margin: 0 auto;
  max-width: 90vw;
  overflow-x: auto;
  position: relative;
`;

const Item = styled.div`
  flex: 0 0 auto;

  @media (min-width: 768px) {
    width: 20%;
  }

  img {
    width: 100%;
    display: block;
  }
`;

const Header = styled.h1``;

const Controls = styled.div`
  width: 90vw;
  margin: 0 auto;

  @media (min-width: 768px) {
    display: flex;
    justify-content: center;
  }

  button {
    background: transparent;
    border: none;
    color: #fff;
    fill: #fff;
    margin-right: 5px;
    margin-top: -25px;
    padding: 5px 10px;
    position: absolute;
    top: 50%;

    + button {
      margin-right: 0;
    }

    @media (min-width: 768px) {
      background: blue;
      margin-top: 0;
    }

    &:focus {
      outline: 0;
    }

    @media (min-width: 768px) {
      position: relative;
    }

    span {
      display: none;

      @media (min-width: 768px) {
        display: block;
      }
    }

    svg {
      height: 50px;

      @media (min-width: 768px) {
        display: none;
      }
    }

    &.prev {
      left: 10px;
      border-radius: 5px 0 0 5px;

      svg {
        transform: rotate(180deg);
      }

      @media (min-width: 768px) {
        left: 0;
      }
    }

    &.next {
      right: 10px;
      border-radius: 0 5px 5px 0;

      @media (min-width: 768px) {
        right: 0;
      }
    }
  }
`;

class Carousel extends Component {
  constructor() {
    super();
    this.state = {
      activeIndex: 0
    };
    this.showNextSet = this.showNextSet.bind(this);
    this.showPrevSet = this.showPrevSet.bind(this);
  }

  showNextSet() {
    if (this.state.activeIndex < 5) {
      this.setState({ activeIndex: this.state.activeIndex + 1 });
    } else {
      this.setState({ activeIndex: 0 });
    }
  }

  showPrevSet() {
    if (this.state.activeIndex > 0) {
      this.setState({ activeIndex: this.state.activeIndex - 1 });
    } else {
      this.setState({ activeIndex: 5 });
    }
  }

  render() {
    const data = this.props.data;
    let sliderStyle = {
      transform: `translateX(${this.state.activeIndex * -100}%)`,
      transition: "0.5s"
    };

    if (data) {
      var projects = data.map(function(projects) {
        return (
          <Item key={projects.id} style={sliderStyle}>
            <img alt={projects.id} src={projects.webformatURL} />
            <h3>{projects.user}</h3>
          </Item>
        );
      });
    }

    return (
      <Section>
        <Header>Carousel Test</Header>
        <Article>{projects}</Article>
        <Controls>
          <button className="carousel__button prev" onClick={this.showPrevSet}>
            <NavigationIcon />
            <span>Prev</span>
          </button>
          <button className="carousel__button next" onClick={this.showNextSet}>
            <span>Next</span>
            <NavigationIcon />
          </button>
        </Controls>
      </Section>
    );
  }
}

export default Carousel;
