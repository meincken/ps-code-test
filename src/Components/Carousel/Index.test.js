import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Index";
import "@testing-library/jest-dom/extend-expect";

test("Check for Title", () => {
  const { getByText } = render(<Carousel />);
  const getTitle = getByText("Carousel Test");
  expect(getTitle).toBeInTheDocument();
});

test("Check for Next Button", () => {
  const { getByText } = render(<Carousel />);
  const getTitle = getByText("Next");
  expect(getTitle).toBeInTheDocument();
});

test("Click Next Button", () => {
  const { getByText } = render(<Carousel />);

  fireEvent(
    getByText("Next"),
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true
    })
  );
});

test("Check for Prev Button", () => {
  const { getByText } = render(<Carousel />);
  const getTitle = getByText("Prev");
  expect(getTitle).toBeInTheDocument();

  fireEvent(
    getByText("Prev"),
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true
    })
  );
});

test("Click Prev Button", () => {
  const { getByText } = render(<Carousel />);

  fireEvent(
    getByText("Prev"),
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true
    })
  );
});
