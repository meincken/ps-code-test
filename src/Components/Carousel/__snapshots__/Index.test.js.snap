// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-AxjAm fzqnPh"
  >
    <h1
      class="sc-AxhCb ccMvAG"
    >
      Carousel Test
    </h1>
    <article
      class="sc-AxirZ icYjrT"
    />
    <div
      class="sc-AxhUy dZUOZv"
    >
      <button
        class="carousel__button prev"
      >
        <svg>
          arrow.svg
        </svg>
        <span>
          Prev
        </span>
      </button>
      <button
        class="carousel__button next"
      >
        <span>
          Next
        </span>
        <svg>
          arrow.svg
        </svg>
      </button>
    </div>
  </section>
</DocumentFragment>
`;
